src/ttlv/fast_scan.rs

```diff
@@ -0,0 +1,1009 @@
+            return Err(FastScanError::assert());
+        }
+
+        let ptr = input.as_ptr().cast::<[u8; 8]>();
+        let input = unsafe { core::slice::from_raw_parts(ptr, num_blocks) };
+        Ok(Self::from_blocks(input))
+    }
+
+    /// Construct a new [`FastScanner`] from a block slice.
+    pub const fn from_blocks(input: &'a [[u8; 8]]) -> Self {
+        Self {
+            remaining: input,
+            failure_word: 0,
+        }
+    }
+
+    /// The remaining input to the scanner.
+    pub const fn remaining(&self) -> &'a [[u8; 8]] {
+        self.remaining
+    }
+
+    /// Whether the scanner is finished.
+    pub const fn is_empty(&self) -> bool {
+        self.remaining.is_empty()
+    }
+
+    /// Whether an error has been delayed.
+    ///
+    /// This should only be checked if the consistency of the previously
+    /// scanned data is important, e.g. if acting on it.  It can be ignored
+    /// while parsing.
+    pub const fn delayed_error(&self) -> bool {
+        self.failure_word != 0
+    }
+
+    /// Whether a whole element is available.
+    ///
+    /// The length of the next element will be checked; if it is entirely loaded
+    /// in the buffer, `true` is returned.  Otherwise, or if there is no next
+    /// element, `false` is returned.
+    pub fn have_next(&self) -> bool {
+        let [head, rest @ ..] = self.remaining else {
+            return false;
+        };
+
+        let length = u32::from_be_bytes(head[4..8].try_into().unwrap());
+        let available = rest.as_flattened().len();
+
+        if usize::MAX as u32 == u32::MAX {
+            // 'usize' is at least as big as 'u32'.
+            available >= length as usize
+        } else {
+            // 'u32' is bigger than 'usize'.
+            available as u32 >= length
+        }
+    }
+
+    /// Finish the scanner.
+    ///
+    /// If there is remaining input, or an error was delayed, [`Err`] is
+    /// returned.  If additional remaining input is allowed, use
+    /// [`Self::finish_non_exhaustive()`] instead.
+    pub fn finish(self) -> Result<(), FastScanError> {
+        if self.is_empty() && !self.delayed_error() {
+            Ok(())
+        } else {
+            Err(FastScanError::assert())
+        }
+    }
+
+    /// Finish the scanner non-exhaustively.
+    ///
+    /// Additional input is ignored.  [`Err`] is returned only if an error is
+    /// delayed.  If additional remaining input is not allowed, use
+    /// [`Self::finish()`] instead.
+    pub fn finish_non_exhaustive(self) -> Result<(), FastScanError> {
+        if !self.delayed_error() {
+            Ok(())
+        } else {
+            Err(FastScanError::assert())
+        }
+    }
+}
+
+impl<'a> FastScanner<'a> {
+    /// Scan an expected structure.
+    ///
+    /// The returned sub-scanner can be used to scan the fields of the
+    /// `struct`; `self` will independently continue from the end of the
+    /// `struct`.
```

Comment:
As all `scan` functions return a sub-scanner and leave `self` at the next position in the input after the encountered TTLV item, maybe say it once at the top somewhere instead of repeating it for each function?